src/protocol/aggregation/mod.rs

```diff
@@ -26,110 +27,124 @@ use crate::{
 
 #[step]
 pub(crate) enum Step {
-    BinaryValidator,
-    UpgradeValueBits,
-    UpgradeBreakdownKeyBits,
+    Validator,
+    ConvertValueBits,
 }
 
 /// Binary-share aggregation protocol.
 ///
 /// # Errors
 /// Propagates errors from multiplications
-pub async fn aggregate<'a, C, SB, F, V, BK>(
+pub async fn aggregate<'a, C, S, SB, F, CV, BK>(
     sh_ctx: C,
-    input_rows: &[AggregateInputRow<V, BK>],
+    input_rows: &[AggregateInputRow<CV, BK>],
 ) -> Result<Vec<Replicated<F>>, Error>
 where
     C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F>
+        + BasicProtocols<C::UpgradedContext<F>, F>
+        + Reshare<C::UpgradedContext<F>, RecordId>
+        + Serializable
+        + DowngradeMalicious<Target = Replicated<F>>
+        + 'static,
     C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB>,
     SB: LinearSecretSharing<Gf2>
         + BasicProtocols<C::UpgradedContext<Gf2>, Gf2>
         + DowngradeMalicious<Target = Replicated<Gf2>>
         + 'static,
     F: PrimeField + ExtendableField,
-    V: GaloisField,
+    CV: GaloisField,
     BK: GaloisField,
 {
-    let validator = sh_ctx.narrow(&Step::BinaryValidator).validator();
-    let m_ctx = validator.context();
+    let validator = sh_ctx.narrow(&Step::Validator).validator::<F>();
 
-    let (gf2_value_bits, gf2_breakdown_keys) = (
+    let (gf2_value_bits, _gf2_breakdown_keys) = (
         get_gf2_value_bits(input_rows),
         get_gf2_breakdown_key_bits(input_rows),
     );
-    let (upgraded_gf2_value_bits, upgraded_gf2_breakdown_key_bits) = try_join(
-        m_ctx
-            .narrow(&Step::UpgradeValueBits)
-            .upgrade(gf2_value_bits),
-        m_ctx
-            .narrow(&Step::UpgradeBreakdownKeyBits)
-            .upgrade(gf2_breakdown_keys),
-    )
-    .await?;
-    let binary_shared_values = zip(upgraded_gf2_value_bits, upgraded_gf2_breakdown_key_bits)
-        .map(|(value, breakdown_key)| BinarySharedAggregateInputs::new(value, breakdown_key))
-        .collect::<Vec<_>>();
-
-    secure_aggregation(validator, binary_shared_values).await
+
+    // TODO(taikiy):
+    // 1. slice the buckets into N streams and send them to the aggregation protocol
+    // 2. collect the results and return them
+
+    let output = aggregate_values(validator.context(), gf2_value_bits).await?;
+
+    validator.validate(vec![output]).await
 }
 
 /// Performs a set of aggregation protocols on binary shared values.
+/// This protocol assumes that the input rows are from the same match key and
+/// for the same breakdown key, so that we don't have to worry about computing
+/// the helper_bits or capping.
```

Comment:
We are not assuming they are all from the same "match key". There are no match keys. We aren't assuming it is from the same device either. We are assuming devices / browsers are properly applying per-user capping to protect people who are using legitimate software.